gtsam/linear/AcceleratedPowerMethod.h

```diff
@@ -0,0 +1,133 @@
+
+using Sparse = Eigen::SparseMatrix<double>;
+
+/* ************************************************************************* */
+/// MINIMUM EIGENVALUE COMPUTATIONS
+
+// Template argument Operator just needs multiplication operator
+template <class Operator>
+class AcceleratedPowerMethod : public PowerMethod<Operator> {
+  /**
+   * \brief Compute maximum Eigenpair with power method
+   *
+   * References :
+   * 1) Rosen, D. and Carlone, L., 2017, September. Computational
+   * enhancements for certifiably correct SLAM. In Proceedings of the
+   * International Conference on Intelligent Robots and Systems.
+   * 2) Yulun Tian and Kasra Khosoussi and David M. Rosen and Jonathan P. How,
+   * 2020, Aug, Distributed Certifiably Correct Pose-Graph Optimization, Arxiv
+   * 3) C. de Sa, B. He, I. Mitliagkas, C. Ré, and P. Xu, “Accelerated
+   * stochastic power iteration,” in Proc. Mach. Learn. Res., no. 84, 2018, pp.
+   * 58–67
+   *
+   * It performs the following iteration: \f$ x_{k+1} = A * x_k + \beta *
+   * x_{k-1} \f$ where A is the certificate matrix, x is the Ritz vector
+   *
+   */
+
+  double beta_ = 0;  // a Polyak momentum term
+
+  Vector previousVector_;  // store previous vector
+
+ public:
+  // Constructor from aim matrix A (given as Matrix or Sparse), optional intial
+  // vector as ritzVector
+  explicit AcceleratedPowerMethod(
+      const Operator &A, const boost::optional<Vector> initial = boost::none)
+      : PowerMethod<Operator>(A, initial) {
+    Vector x0;
+    // initialize ritz vector
+    x0 = initial ? Vector::Random(this->dim_) : initial.get();
+    Vector x00 = Vector::Random(this->dim_);
+    x0.normalize();
+    x00.normalize();
+
+    // initialize Ritz eigen vector and previous vector
+    previousVector_ = update(x0, x00, beta_);
+    this->ritzVector_ = update(previousVector_, x0, beta_);
+    this->perturb();
+
+    // set beta
+    Vector init_resid = this->ritzVector_;
+    const double up = init_resid.transpose() * this->A_ * init_resid;
+    const double down = init_resid.transpose().dot(init_resid);
+    const double mu = up / down;
+    beta_ = mu * mu / 4;
+    setBeta();
+  }
+
+  // Update the ritzVector with beta and previous two ritzVector
+  Vector update(const Vector &x1, const Vector &x0, const double beta) const {
+    Vector y = this->A_ * x1 - beta * x0;
+    y.normalize();
+    return y;
+  }
+
+  // Update the ritzVector with beta and previous two ritzVector
+  Vector update() const {
+    Vector y = update(this->ritzVector_, previousVector_, beta_);
+    previousVector_ = this->ritzVector_;
+    return y;
+  }
+
+  // Tuning the momentum beta using the Best Heavy Ball algorithm in Ref(3)
+  void setBeta() {
+    double maxBeta = beta_;
+    size_t maxIndex;
+    std::vector<double> betas = {2 / 3 * maxBeta, 0.99 * maxBeta, maxBeta,
+                                 1.01 * maxBeta, 1.5 * maxBeta};
+
+    Matrix R = Matrix::Zero(this->dim_, 10);
+    for (size_t i = 0; i < 10; i++) {
```

Comment:
This is the *number of Iterations T* in the Best Heavy Ball algorithm. It means we need to do T times iteration to find the initial ritz vector that returns the largest Rayleigh quotient (But here I didn't randomly initialize this vector, have corrected this).

The paper did not mention how to set the T, so I use 10 here. But to be honest, I didn't find information regarding how to set this iteration time. Do you have idea? Or I think we can just skip the outer iteration since we've already have a `w` vector in the algorithm, which is also the initial ritz vector.

The algorithm is the B.3 in this [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6557638/)

Reply:
Yes. I should follow this naming style.